server/accounts.go

```diff
@@ -490,6 +494,236 @@ func (a *Account) TotalSubs() int {
 	return int(a.sl.Count())
 }
 
+// MapDest is for mapping published subjects for clients.
+type MapDest struct {
+	Subject string
+	Weight  uint8
+}
+
+// destination is for internal representation for a weighted mapped destination.
+type destination struct {
+	tr     *transform
+	weight uint8
+}
+
+// mapping is an internal entry for mapping subjects.
+type mapping struct {
+	src   string
+	wc    bool
+	dests []*destination
+}
+
+// AddMapping adds in a simple route mapping from src subject to dest subject
+// for inbound client messages.
+func (a *Account) AddMapping(src, dest string) error {
+	return a.AddWeightedMappings(src, &MapDest{dest, 100})
+}
+
+// AddWeightedMapping will add in a weighted mappings for the destinations.
+// TODO(dlc) - Allow cluster filtering
+func (a *Account) AddWeightedMappings(src string, dests ...*MapDest) error {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	// We use this for selecting between multiple weighted destinations.
+	if a.prand == nil {
+		a.prand = rand.New(rand.NewSource(time.Now().UnixNano()))
+	}
+
+	if !IsValidSubject(src) {
+		return ErrBadSubject
+	}
+
+	m := &mapping{src: src, wc: subjectHasWildcard(src), dests: make([]*destination, 0, len(dests)+1)}
+	seen := make(map[string]struct{})
+
+	var tw uint8
+	for _, d := range dests {
+		if _, ok := seen[d.Subject]; ok {
+			return fmt.Errorf("duplicate entry for %q", d.Subject)
+		}
+		seen[d.Subject] = struct{}{}
+		if d.Weight > 100 {
+			return fmt.Errorf("individual weights need to be <= 100")
+		}
+		tw += d.Weight
+		if tw > 100 {
+			return fmt.Errorf("total weight needs to be <= 100")
+		}
+		if !IsValidSubject(d.Subject) {
+			return ErrBadSubject
+		}
+		tr, err := newTransform(src, d.Subject)
+		if err != nil {
+			return err
+		}
+
+		m.dests = append(m.dests, &destination{tr, d.Weight})
+	}
+
+	// Auto add in original at weight if all entries weight does not total to 100.
```

Comment:
I'd have a special one so users can easily inject loss for testing.
And if not, I'd force the sum to always be 100, be explicit

Reply:
Not following, I coded explicit, was not good always doing math especially if say you have a service and you just want some traffic to be diverted.

If you have a destination that no one subscribes to that will be loss ;)

Reply:
Say I have a subscriber foo.
For testing a user application behaving under loss.
I set the weighted transform to 90% or -10% and we drop 10 percent of messages. With -0
Thought that could be useful, which is why I brought it up

Reply:
Ah so do not force the sum to be 100 to allow for loss. ok I think I understand now.

Reply:
Would `foo -> {foo:80}` work? Meaning that this would mean we want 20% loss?

Reply:
If so we can add the behavior in later since the interface would be same.

Reply:
That would be the idea yes. 

I wouldn't wait too long, just so we can still control what the behavior should be if the values do not add up to 100%

Reply:
ok let me add in the behavior now.